src/coreclr/src/vm/nativeimage.cpp

```diff
@@ -0,0 +1,150 @@
+
+#include <shlwapi.h>
+
+AssemblyNameIndex::AssemblyNameIndex(const SString& name, int32_t index)
+    : Name(name), Index(index)
+{
+    Name.Normalize();
+}
+
+NativeImage::NativeImage(
+    PEFile *pPeFile,
+    PEImageLayout *pPeImageLayout,
+    READYTORUN_HEADER *pHeader,
+    LPCUTF8 nativeImageName,
+    LoaderAllocator *pLoaderAllocator,
+    AllocMemTracker& amTracker)
+    : m_eagerFixupsLock(CrstNativeImageEagerFixups)
+{
+    CONTRACTL
+    {
+        CONSTRUCTOR_CHECK;
+        STANDARD_VM_CHECK;
+        INJECT_FAULT(COMPlusThrowOM(););
+    }
+    CONTRACTL_END;
+
+    LoaderHeap *pHeap = pLoaderAllocator->GetHighFrequencyHeap();
+    m_utf8SimpleName = nativeImageName;
+    m_pPeImageLayout = pPeImageLayout;
+    m_eagerFixupsHaveRun = false;
+    
+    m_pReadyToRunInfo.Assign(new (amTracker.Track(pHeap->AllocMem((S_SIZE_T)sizeof(ReadyToRunInfo))))
+        ReadyToRunInfo(/*pModule*/ NULL, pPeImageLayout, pHeader, /*compositeImage*/ NULL, &amTracker), /*takeOwnership*/ TRUE);
+    m_pComponentAssemblies = m_pReadyToRunInfo->FindSection(ReadyToRunSectionType::ComponentAssemblies);
+    m_componentAssemblyCount = m_pComponentAssemblies->Size / sizeof(READYTORUN_COMPONENT_ASSEMBLIES_ENTRY);
+    
+    // Check if the current module's image has native manifest metadata, otherwise the current->GetNativeAssemblyImport() asserts.
+    m_pManifestMetadata = LoadManifestMetadata();
+
+    HENUMInternal assemblyEnum;
+    HRESULT hr = m_pManifestMetadata->EnumAllInit(mdtAssemblyRef, &assemblyEnum);
+    mdAssemblyRef assemblyRef;
+    int assemblyIndex = 0;
+    while (m_pManifestMetadata->EnumNext(&assemblyEnum, &assemblyRef))
+    {
+        LPCSTR assemblyName;
+        hr = m_pManifestMetadata->GetAssemblyRefProps(assemblyRef, NULL, NULL, &assemblyName, NULL, NULL, NULL, NULL);
+        m_assemblySimpleNameToIndexMap.Add(new (amTracker.Track(pHeap->AllocMem((S_SIZE_T)sizeof(AssemblyNameIndex))))
+            AssemblyNameIndex(SString(SString::Ansi, assemblyName), assemblyIndex));
+        assemblyIndex++;
+    }
+}
+
+NativeImage::~NativeImage()
+{
+    if (m_pManifestMetadata != NULL)
+    {
+        m_pManifestMetadata->Release();
+    }
+}
+
+bool NativeImage::Matches(LPCUTF8 utf8SimpleName) const
+{
+    return !_stricmp(utf8SimpleName, m_utf8SimpleName);
+}
+
+NativeImage *NativeImage::Open(
+    PEFile *pPeFile,
+    PEImageLayout *pPeImageLayout,
+    LPCUTF8 nativeImageName,
+    LoaderAllocator *pLoaderAllocator)
+{
+#ifndef DACCESS_COMPILE
+    uint32_t headerRVA = pPeImageLayout->GetExport("RTR_HEADER");
+    if (headerRVA + sizeof(READYTORUN_HEADER) > pPeImageLayout->GetSize())
+    {
+        return NULL;
+    }
+    READYTORUN_HEADER *pHeader = (READYTORUN_HEADER *)((BYTE *)pPeImageLayout->GetBase() + headerRVA);
+    AllocMemTracker amTracker;
+    NativeImage *result = new NativeImage(pPeFile, pPeImageLayout, pHeader, nativeImageName, pLoaderAllocator, amTracker);
+    amTracker.SuppressRelease();
```

Comment:
If you would like to keep using LoaderHeap, SuppressRelease should be done only after the structure is registered with global state - that would be after `m_nativeImages.Append(nativeImage);`.

Reply:
Hmm, now I'm somewhat confused - ReadyToRunInfo ctor does need an AllocMemTracker argument and the "legacy" ReadyToRunInfo::Initialize does initialize the ReadyToRunInfo using the LoaderHeap and AllocMemTracker. I was originally trying to follow the same pattern but if it's not the right pattern to use, I could probably use some advice as to what is the preferable pattern here.

Reply:
Ah ok. It is fine to keep allocating some of this on LoaderHeap with AllocMemTracker to follow existing pattern.

Reply:
Moved AllocMemTracker to AssemblyLoadContext::LoadNativeImage in 14th commit per your suggestion.